src/helpers/buffers/fsv.rs

```diff
@@ -0,0 +1,220 @@
+///
+/// Elements can be added to this vector in random order as long as the total number of elements
+/// does not exceed the overall capacity of the vector. An attempt to do so, will lead to a panic.
+/// Every element is stored by a specific offset and it is assumed that all elements are of the same
+/// size.
+///
+/// The layout of the vector with some elements added to it is presented below. In this example,
+/// vector has 3 regions and `X` indicates that space at that element is occupied.
+///
+///  region1  region2  region3
+/// [X,_,_,_][_,_,X,_][_,X,_,X]
+///
+/// Once `region1` is completely filled up, it is possible to drain the vector
+///
+/// [X,X,X,X][_,_,X,_][_,X,_,X] -> `drain` -> [_,_,X,_][_,X,_,X][_,_,_,_]
+///
+/// This vector is used inside the send buffer to keep track of messages added to it. Once first
+/// batch of messages is ready (region1 is full), it drains this vector and send those messages
+/// down to the network layer
+#[derive(Debug)]
+pub struct FixedSizeByteVec<const N: usize> {
+    data: Vec<u8>,
+    region_size: usize,
+    added: BitVec,
+    elements_drained: usize,
+}
+
+impl<const N: usize> FixedSizeByteVec<N> {
+    pub const ELEMENT_SIZE_BYTES: usize = N;
+
+    pub fn new(region_count: usize, region_size: usize) -> Self {
+        Self {
+            data: vec![0_u8; N * region_size * region_count],
+            added: bitvec![0; region_size * region_count],
+            elements_drained: 0,
+            region_size,
+        }
+    }
+
+    /// Inserts a new element to the specified position, returning the previous element at this `index`.
+    /// ## Panics
+    /// * Panics if `index` is out of bounds.
+    pub fn insert(&mut self, index: usize, elem: [u8; N]) -> Option<[u8; N]> {
+        let offset = Self::index_offset(index);
+
+        // if index is out of bounds, this line will panic, there is no need for additional check
+        if self.added[index] {
+            // this is not supposed to be on the hot path, this branch is executed when there is
+            // an error and message with the same record id is received more than once, so
+            // copying here is fine.
+            let r = self.data[offset.clone()].try_into().unwrap();
+            self.data[offset].copy_from_slice(&elem);
+
+            Some(r)
+        } else {
+            self.added.set(index, true);
+            self.data[offset].copy_from_slice(&elem);
+
+            None
+        }
+    }
+
+    /// Checks whether the first region of elements is ready to be drained.
+    pub fn ready(&self) -> bool {
+        self.added[..self.region_size].all()
+    }
+
+    /// Drains the elements from the beginning of the vector. If all of the first `self.region_size`
+    /// elements have been inserted, returns them in order. `self.ready()` should indicate that.
+    ///
+    /// ## Panic
+    /// Panics if `ready()` is false before calling `drain()`.
+    pub fn drain(&mut self) -> Vec<u8> {
```

Comment:
Maybe you could write this as `take(&mut self) -> Option<Vec<u8>>` and avoid separate ready and drain calls at the caller side.  Then you don't need to worry about assertions in this one.